File: stacks/reversePolishNotation/reversePolishNotation.py
# O(n) time | O(n) space


def reversePolishNotation(tokens):
    stack = []

    for token in tokens:
        if token == "+":
            stack.append(stack.pop() + stack.pop())
        elif token == "-":
            firstNumber = stack.pop()
            stack.append(stack.pop() - firstNumber)
        elif token == "*":
            stack.append(stack.pop() * stack.pop())
        elif token == "/":
            firstNumber = stack.pop()
            stack.append(int(stack.pop() / firstNumber))
        else:
            stack.append(int(token))

    return stack.pop()


# Example usage:

print(reversePolishNotation(["2", "1", "+", "3", "*"]))
# Output: 9

File: stacks/reversePolishNotation/reversePolishNotation.ts
// O(n) time | O(n) space

function reversePolishNotation(tokens: string[]) {

  const stack: number[] = []
  for (const token of tokens) {
    if(token === "+"){
      stack.push(stack.pop()! + stack.pop()!)
    } 
    else if(token === "-"){
      const firstNum = stack.pop()!
      stack.push(stack.pop()! - firstNum)
    } 
    else if(token === "*"){
      stack.push(stack.pop()! * stack.pop()!)
    } 
    else if(token === "/"){
      const firstNum = stack.pop()!
      stack.push(Math.trunc(stack.pop()! / firstNum))
    } else {
      stack.push(Number(token))
    }
  }

  return stack.pop()!
  
}

// Example usage:
const tokens = ["2", "1", "+", "3", "*"];
console.log(reversePolishNotation(tokens)); // Output: 9
const tokens2 = ["4", "13", "5", "/", "+"];
console.log(reversePolishNotation(tokens2)); // Output: 6
